Add an option to disable the creation of the forward index for a column

somandal opened the pull request:
This PR adds an option to disable the forward index for a given column via the FieldConfig properties list. This is a PR to solve issue https://github.com/apache/pinot/issues/6473. 

If the forward index is disabled, only a subset of queries can work. Basically queries that don't need to select, transform, group by, order by but do need to use the column in WHERE clauses can be supported.

To disable the forward index, the following mandates have been added at the moment (depending on the usage of this feature we may decide to relax some of them over time):

- The inverted index for the column must be enabled
    - We require this to add functionality on the reload path to regenerate the forward index in case users decide later that they want a forward index
- The column must be a dictionary based column
    - We require this to add functionality on the reload path to regenerate the forward index in case users decide later that they want a forward index
- The column cannot have a V1 type range index present or be a multi-value columns with a range index present
    - V1 type range index resorts to full scans and needs the forward index to work properly
    - Multi-value columns always use the V1 type range index

The above checks have been added in the code in the following places:

- TableConfig validator
- Segment reload path (check when creating the default column)
- Segment creation path


Sorted Column Handling:

For sorted columns we decided to allow someone to disable the forward index, but internally it's a no-op operation (we already create only a single index for sorted columns and use it for both forward and inverted index, we'll continue doing the same irrespective of the `forwardIndexDisabled` flag)


The ability to disable the forward index is currently implemented for the following cases:

- New immutable segments generated for a column (all refresh use cases and reload use cases for new segments)

This PR does **not** add support for disabling the forward index for:

- Mutable segments
- Handling default and derived columns with the forward index disabled. We have decided to address default columns in the following way (derived still needs thought):
    - SV default columns should continue to create a sorted forward index (similar to the segment creation code path). We are okay with query inconsistency behavior for end users in this scenario.
    - MV default columns should not have a forward index but should have the inverted index + dictionary and any other indexes enabled during reload.
- Enabling / disabling the disable forward index flag for an existing column -> we will add support for this reload case in a follow up PR shortly after this PR


Validations with other indices can be broadly classified into the following paths:

- Query path - ensure that none of the other indices rely on the forward index
    - We found the range index v1 resorts to a scan on the forward index, thus disabled this feature if a range index of version 1 is present
    - Sorted columns rely on the forward index as the inverted index as well. We treat sorted column as a no-op when this feature is enabled (do not throw an error)
- Segment creation path does not rely on the forward index to create other indices. The `indexRow` function creates all the indices needed at the same time for a given row.
- Segment reload path relies on the forward index to add new indices such as range index, text index, etc. This path cannot be supported until we address enabling/disabling this feature on the reload path and will be done as part of a future PR along with any additional validations needed.


No-op creators have been added for the forward index disabled columns. This is useful for:
- On the segment creation side, we can keep code clean and reuse the forward index creator map to walk through all the columns and create other indices


All code paths that need to read the forward index now have null checks added and appropriate exceptions thrown if the forward index doesn't exist.


cc @siddharthteotia @Jackie-Jiang

somandal commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/IndexCreationContext.java:
```diff
@@ -77,6 +79,7 @@ final class Builder {
     private boolean _hasDictionary = true;
     private Comparable<?> _minValue;
     private Comparable<?> _maxValue;
+    private boolean _forwardIndexDisabled = false;
```
good catch! done

somandal commented on pinot-segment-local/src/test/resources/data/newColumnsSchemaWithForwardIndexDisabled.json:
```diff
@@ -0,0 +1,77 @@
+      "name": "daysSinceEpoch"
+    }
+  }
+}
```
done

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/invertedindex/H3IndexHandler.java:
```diff
@@ -155,18 +155,30 @@ private void handleDictionaryBasedColumn(SegmentDirectory.Writer segmentWriter,
+        // Create the h3 index if the dictionary length is 1 as this is for a default column (i.e. newly added
+        // column). For existing columns it is not possible to create the h3 index without forward index
+        Preconditions.checkState(dictionary.length() == 1,
+            "Creating h3 index for forward index disabled default column, dictionary size must be 1");
```
good catch, updated here and in other related IndexHandler files

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/datasource/MutableDataSource.java:
```diff
@@ -106,6 +106,12 @@ public int getMaxNumValuesPerMVEntry() {
 
+    @Override
+    public boolean forwardIndexDisabled() {
+      // Disabling the forward index is only supported for offline segments.
```
yes meant immutable segments, done

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -196,6 +204,14 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
+        Preconditions.checkState(dictEnabledColumn && invertedIndexColumns.contains(columnName)
+            && !columnIndexCreationInfo.isSorted() && (!rangeIndexColumns.contains(columnName)
+                || rangeIndexVersion == BitSlicedRangeIndexCreator.VERSION), "Cannot disable forward index for column "
+            + "without dictionary and inverted index or which is sorted or which has range index with version < 2");
```
done

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -196,6 +204,15 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
+            && !columnIndexCreationInfo.isSorted() && (!rangeIndexColumns.contains(columnName)
+                || rangeIndexVersion == BitSlicedRangeIndexCreator.VERSION), String.format("Cannot disable forward "
+            + "index for column: %s without dictionary and inverted index or which is sorted or which has range "
+            + "index with version < 2", columnName));
```
(nit) may be cleaner to split this into 2-3 conditions. Will also make it clear to user what exact combination is invalid. Consider writing a simple helper method and fold everything into it

somandal replied:
good point, done!

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/MultiValueNoOpForwardIndexCreator.java:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  @Override
+  public void putDictIdMV(int[] dictIds) {
```
Similar to NO-OP reader, do we need to throw specific UNSUP here ?

siddharthteotia replied:
Nvm.. I now remember our offline discussion and these empty methods are needed to keep the segment generation code clean

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/SingleValueNoOpForwardIndexCreator.java:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  @Override
+  public void putDictId(int dictId) {
```
Same here

siddharthteotia replied:
Ignore

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java:
```diff
@@ -142,6 +145,8 @@ public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, Colum
         // Single-value
         if (metadata.isSorted()) {
           // Sorted
+          // forwardIndexDisabled columns do not need to be handled here as forward index cannot be disabled on a
```
May be I am mis-reading the if-else block here. For an unsorted dict column with inv index, are we ok with line 157 being executed with null `fwdIndexBuffer` ?

somandal replied:
yes it is okay for 157 to be called with `fwdIndexBuffer` as null. We only set this to null if the forward index is disabled. When the forward index reader is created, the No-op forward index reader will be assigned which doesn't need a `fwdIndexBuffer`. I've added a comment where I set `fwdIndexBuffer` to indicate this behavior is expected. Let me know if you would like more clarification on this.

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/DefaultIndexCreatorProvider.java:
```diff
@@ -88,17 +90,25 @@ public ForwardIndexCreator newForwardIndexCreator(IndexCreationContext.Forward c
-        if (context.isSorted()) {
-          return new SingleValueSortedForwardIndexCreator(context.getIndexDir(), context.getFieldSpec().getName(),
-              context.getCardinality());
+      if (context.forwardIndexDisabled()) {
```
(nit) not related to this PR -- do you mind adding a comment under the else branch above line 93 ? something like `Dictionary enabled column`. Will make it bit more readable. Similarly above line 79.

We may also want to make this more readable by having separate helper method but separate PR for that is fine too

somandal replied:
done, will do the helper method changes as a follow up to avoid confusion in this PR. :)

siddharthteotia commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java:
```diff
@@ -143,7 +143,7 @@ public class OfflineClusterIntegrationTest extends BaseClusterIntegrationTestSet
   // TODO: This might lead to flaky test, as this disk size is not deterministic
   //       as it depends on the iteration order of a HashSet.
-  private static final int DISK_SIZE_IN_BYTES = 20797128;
+  private static final int DISK_SIZE_IN_BYTES = 20843088;
```
Why this change ?

somandal replied:
so I've added a metadata field in the column metadata which increases the disk size. Also based on the comment here it looks like the size may change depending on the iteration order of the `HashSet`. Let me know if you think I should look into this deeper to compare the difference in size change.

somandal replied:
This change is no longer needed since we removed the field from the metadata. I've changed it back to the original value.

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:
```diff
@@ -869,6 +871,25 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
             break;
         }
+
+        // Validate the forward index disabled compatibility if enabled for this column
```
(nit) put this in a helper method ?

somandal replied:
done

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:
```diff
@@ -869,6 +871,25 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
+          boolean forwardIndexDisabled = Boolean.parseBoolean(fieldConfigProperties
+              .getOrDefault(FieldConfig.FORWARD_INDEX_DISABLED, FieldConfig.DEFAULT_FORWARD_INDEX_DISABLED));
+          if (forwardIndexDisabled) {
+            Preconditions.checkState((noDictionaryColumns == null || !noDictionaryColumns.contains(columnName))
```
done

Jackie-Jiang commented:
We should allow disabling forward index for sorted column, which is a no-op because the inverted index for sorted column can be used as forward index. I believe it is already allowed in the implementation because the default column is always sorted. No need to add this in the validation.

somandal commented:
> We should allow disabling forward index for sorted column, which is a no-op because the inverted index for sorted column can be used as forward index. I believe it is already allowed in the implementation because the default column is always sorted. No need to add this in the validation.

Thanks for the suggestion @Jackie-Jiang. From going through the code it looks like for sorted columns the forward index and inverted index are essentially the same. Code snippet from the `PhysicalColumnIndexCreator`:

```
        // Single-value
        if (metadata.isSorted()) {
          // Sorted
          // forwardIndexDisabled columns do not need to be handled here as forward index cannot be disabled on a
          // sorted column
          SortedIndexReader<?> sortedIndexReader = indexReaderProvider.newSortedIndexReader(fwdIndexBuffer, metadata);
          _forwardIndex = sortedIndexReader;
          _invertedIndex = sortedIndexReader;
          _fstIndex = null;
          return;
        }
```

I also see code on the segment creation path where we skip creating the inverted index if the column is sorted even if the inverted index is enabled for the column. So keeping these in mind, we didn't think it made much sense to disable forward index for such columns.

Let me know your thoughts based on the above. cc @siddharthteotia

Jackie-Jiang commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/datasource/DataSourceMetadata.java:
```diff
@@ -70,6 +70,11 @@ default boolean isSingleValue() {
+  /**
+   * Returns whether forward index is disabled for this column or not, returns 'true' if disabled
+   */
+  boolean forwardIndexDisabled();
```
I don't think we need this in the metadata. Instead, we can make `DateSource.getForwardIndex()` nullable, and perform proper null checks when it is used. This is consistent with how we handle other indexes.

somandal replied:
i removed this as it isn't used at the moment

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/NoOpMVForwardIndexReader.java:
```diff
@@ -0,0 +1,66 @@
+ * No-op forward index reader meant to be used by multi-value columns with forward index disabled. This is meant to
+ * provide clean exception handling for operations which need to read forward index when forward index is disabled.
+ */
+public class NoOpMVForwardIndexReader implements ForwardIndexReader<ForwardIndexReaderContext> {
```
Suggest not using no-op reader as this is not consistent with how we handle other index types. Instead, make `DataSource.getForwardIndex()` return `null` when the forward index does not exist. When forward index is needed to solve the query, check if the forward index is `null` and throw exception accordingly

siddharthteotia replied:
I feel that will be quite tricky and less clean since that requires policing multiple areas in the code and then throw exception. IIRC, we had considered this option but NO-OP reader looked better / cleaner

siddharthteotia replied:
GROUP BY
Transform
Aggregate
SELECT
SCAN based filter
.....


Whoever accesses forward index now needs to be aware of the non-existence of forward index. The NO-OP reader on the other hand does the exact same thing but fails just slightly later by letting the caller attempting to read forward index and then fail with UNSUP. This is much cleaner imo.

Jackie-Jiang replied:
We should treat forward index the same as other type of indexes, instead of specializing it and always put a dummy reader for it. Let's say in the future I want to support GROUP BY with inverted index only, my branching will be `if (dataSource,getForwardIndex() == null) ...`, similar to when we choose to use the scan based filter when inverted index doesn't exist.
Conceptually we are able to support all operations with inverted index only, so I don't want to specialize forward index just for the exception message.

siddharthteotia replied:
I am not sure I follow. Throwing exception in multiple places using if checks is special casing / specializing it as opposed to using NO-OP reader like done in this PR. We can discuss this more

siddharthteotia replied:
Unless there is a way to do what you are suggesting without leaking it in several different places in the code, I don't think it is a good idea.

somandal replied:
+1 to what @siddharthteotia says about this. There are too many places which will require null checks and even then we might miss some. New code will also need to deal with such null checks.

Jackie-Jiang replied:
Not really. We have less than 10 usages of `DataSource.getForwardIndex()` in the production code, so it should be easy to track each of them and ensure the no-forward-index case is properly handled. I don't like the idea of handling forward index differently from other index types - all other index types return `null` when the index does not exist.

Jackie-Jiang replied:
@siddharthteotia More thoughts on this? Adding a dummy reader can simplify the implementation for now, but in the future we will very likely need to change it to know whether the forward index exist (e.g when supporting re-generating forward index from inverted index). One way to do it is adding it to the data source metadata, but that will make the index handling inconsistent.

somandal replied:
we discussed and agreed to remove the no-op reader. I've made these changes. thanks for your patience to discuss this @Jackie-Jiang and @siddharthteotia !

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -196,6 +204,15 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
+      boolean forwardIndexDisabled = forwardIndexDisabledColumns.contains(columnName);
+      if (forwardIndexDisabled) {
+        Preconditions.checkState(dictEnabledColumn && invertedIndexColumns.contains(columnName)
+            && !columnIndexCreationInfo.isSorted() && (!rangeIndexColumns.contains(columnName)
```
hey @Jackie-Jiang can you check my comment about sorted index on your first comment on this PR? Wanted to know your thoughts based on that.

Jackie-Jiang commented:
@somandal We want to disable the forward index to save the storage. For sorted index, we store the index in inverted format (map from dictId to doc range), and also use it as forward index. We should allow disabling the forward index for it in the table config, and simply ignore it when creating the index.

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/DefaultIndexCreatorProvider.java:
```diff
@@ -88,17 +90,25 @@ public ForwardIndexCreator newForwardIndexCreator(IndexCreationContext.Forward c
-              context.getCardinality());
+      if (context.forwardIndexDisabled()) {
+        if (context.getFieldSpec().isSingleValueField()) {
+          return new SingleValueNoOpForwardIndexCreator();
```
done, consolidated this into 1 class. will address removing this as a future change

somandal commented:
> @somandal We want to disable the forward index to save the storage. For sorted index, we store the index in inverted format (map from dictId to doc range), and also use it as forward index. We should allow disabling the forward index for it in the table config, and simply ignore it when creating the index.

Thanks for discussing this, done!

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -816,6 +835,7 @@ public static void addColumnMetadataInfo(PropertiesConfiguration properties, Str
         String.valueOf(columnIndexCreationInfo.getTotalNumberOfEntries()));
     properties.setProperty(getKeyFor(column, IS_AUTO_GENERATED),
         String.valueOf(columnIndexCreationInfo.isAutoGenerated()));
+    properties.setProperty(getKeyFor(column, FORWARD_INDEX_DISABLED), String.valueOf(forwardIndexDisabled));
```
done! remove this

somandal commented:
@Jackie-Jiang @siddharthteotia @jackjlli I've updated the PR with the latest discussions. Let me know what you think! Thanks!

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/operator/filter/ScanBasedFilterOperator.java:
```diff
@@ -48,10 +49,16 @@ public class ScanBasedFilterOperator extends BaseFilterOperator {
   @Override
   protected FilterBlock getNextBlock() {
     DataSourceMetadata dataSourceMetadata = _dataSource.getDataSourceMetadata();
+    ForwardIndexReader<?> forwardIndexReader = _dataSource.getForwardIndex();
```
Instead of `getNextBlock`(), we can probably fail sooner in the constructor of `ScanBasedFilterOperator` which is still during physical planning ?

somandal replied:
Good catch! moved this to the constructor

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java:
```diff
@@ -133,7 +133,12 @@ public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, Colum
+    // forward index disabled columns which doesn't require the 'fwdIndexBuffer'.
+    boolean forwardIndexDisabled = !segmentReader.hasIndexFor(columnName, ColumnIndexType.FORWARD_INDEX);
+    PinotDataBuffer fwdIndexBuffer = forwardIndexDisabled ? null
+        : segmentReader.getIndexFor(columnName, ColumnIndexType.FORWARD_INDEX);
```
Can we replace 2 calls with just 1 call to `getIndexFor` ?

somandal replied:
I can only do this if I put the `getIndexFor` in a `try catch` block. The underlying implementation of the `ColumnIndexDirectory` throws an exception if the index is not present in the `getIndexFor` call. Do you want me to add this in a `try catch` block like this?:

```
PinotDataBuffer fwdIndexBuffer;
try {
    fwdIndexBuffer = segmentReader.getIndexFor(columnName, ColumnIndexType.FORWARD_INDEX);
} catch (Exception e) {
    fwdIndexBuffer = null;
}
```

e.g. of the type of exception thrown down the line (`SingleFileIndexDirectory`):

```
  private PinotDataBuffer checkAndGetIndexBuffer(String column, ColumnIndexType type) {
    IndexKey key = new IndexKey(column, type);
    IndexEntry entry = _columnEntries.get(key);
    if (entry == null || entry._buffer == null) {
      throw new RuntimeException(
          "Could not find index for column: " + column + ", type: " + type + ", segment: " + _segmentDirectory
              .toString());
    }
    return entry._buffer;
  }
```

siddharthteotia replied:
I see. No need for this I guess. Leave it as is

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java:
```diff
@@ -315,6 +341,14 @@ public List<String> getSortedColumns() {
+   * For tests only.
+   */
+  @VisibleForTesting
+  public void setSortedColumns(List<String> sortedColumns) {
```
(nit) can't be a list

somandal replied:
done

jackjlli commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/converter/stats/MutableColumnStatistics.java:
```diff
@@ -145,6 +146,8 @@ public boolean isSorted() {
 
     // Iterate over all data to figure out whether or not it's in sorted order
     MutableForwardIndex mutableForwardIndex = (MutableForwardIndex) _dataSource.getForwardIndex();
+    Preconditions.checkState(mutableForwardIndex != null,
```
I saw that in some places `UnsupportedOperationException` is thrown and in others the precondition is used. Should we consider unifying them?

somandal replied:
Thanks for the review!

I've done that on purpose. Where I throw `UnsupportedOperationException` are places where the forward index is allowed to be null due to disabling it. The exception is just trying to indicate that the query isn't supported.

On the other hand, we don't allow disabling the forward index for `Mutable` segments at the moment. So I added the `Precondition` more as an assert that the forward index absolutely should exist otherwise it means something is very wrong.

Let me know if you still think these should be unified or if the above reasoning seems alright.

somandal replied:
Jackie had a similar comment so I've gone ahead and unified it. I think the exception message is enough to clarify when it's not supported due to a valid disabled forward index and where the forward index shouldn't be null.

jackjlli commented on pinot-tools/src/main/java/org/apache/pinot/tools/segment/converter/DictionaryToRawIndexConverter.java:
```diff
@@ -296,6 +297,10 @@ private void convertOneColumn(IndexSegment segment, String column, File newSegme
       throws IOException {
     DataSource dataSource = segment.getDataSource(column);
     ForwardIndexReader reader = dataSource.getForwardIndex();
+    if (reader == null) {
```
same here. Consider renaming it to `forwardIndexReader`.

somandal replied:
This is a refactoring change (since the original variable was named `reader`). Is it alright if I do this as a follow up to keep this PR simpler? I can check for this in a bunch of other files too then.

somandal replied:
done

jackjlli reviewed (approved):
Minor comments but LGTM.

somandal commented on pinot-core/src/main/java/org/apache/pinot/core/minion/RawIndexConverter.java:
```diff
@@ -204,6 +204,11 @@ private void convertColumn(FieldSpec fieldSpec)
     // Create the raw index
     DataSource dataSource = _originalImmutableSegment.getDataSource(columnName);
     ForwardIndexReader reader = dataSource.getForwardIndex();
+    if (reader == null) {
```
This is a refactoring change (since the original variable was named `reader`). Is it alright if I do this as a follow up to keep this PR simpler? I can check for this in a bunch of other files too then.

somandal replied:
done

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/invertedindex/TextIndexHandler.java:
```diff
@@ -174,18 +175,46 @@ private void createTextIndexForColumn(SegmentDirectory.Writer segmentWriter, Col
+        try (Dictionary dictionary = LoaderUtils.getDictionary(segmentWriter, columnMetadata)) {
+          // Create the text index if the dictionary length is 1 as this is for a default column (i.e. newly added
+          // column). For existing columns it is not possible to create the text index without forward index
+          Preconditions.checkState(dictionary.length() == 1, String.format("Creating text index for forward index "
```
Suggest not adding this part but directly throwing exception. I don't see the point of special handling generating text index for a column with all same values. Same for other index handlers.

somandal replied:
Hi @Jackie-Jiang what do you mean by directly throwing exception?

Without adding this special handling, creation of these indexes for default columns will fail. Today when a new column is added, a default forward index is created. On the reload path I'm skipping the creation of the forward index when this feature is disabled for such default columns. I still create a dictionary though. I do want to create other indices since queries may fail if such segments are accessed and the expected indexes don't exist, right? Or am I missing something here? (i did see the segment preprocessor tests fail when I didn't do special handling for the inverted index for example since the default handlers directly access the forward index to create the auxiliary index)

Jackie-Jiang replied:
Hmm.. That is unexpected. When the dictionary size is 1, the column should be sorted, and we shouldn't skip the forward index

somandal replied:
So we actually evaluated two approaches for this part:

- Create the forward index, allow it to be marked as sorted. This will skip creation of the inverted index, and all other index handlers will kick in and create the index off of the forward index we create. (some index handlers rely on the forward index to construct the index)
    - Our main concern with this approach was that the segments will be inconsistent, some will have a forward index and others won't. This may give odd behavior at query time depending on the segment touched (e.g. some queries go fine with a given filter but don't go through with a slightly different filter). We wanted to avoid this.
    - We had also explored an option to identify such queries on the Broker or Server planning stage but quickly found that this became very ugly and had lots of exceptions and was error prone.
- Don't create the forward index but do create the dictionary (just like the normal segment creation path). Modify the index handlers to create the index for such a default column if forward index is disabled without relying on the the forward index. In such cases we expect exactly 1 value anyways. This will give a consistent experience to users who run queries.
    - This has the disadvantage of doing some special handling in the index handlers for this scenario since the forward index won't be available anymore.

Hope the above explains our thinking about the approaches and why we decided to go with the second one. Let me know if you'd like to discuss this in more detail. cc @siddharthteotia

Jackie-Jiang replied:
I don't follow the second approach. For the default column where all the values are the same for a column, we should create sorted index and a single entry dictionary. This is the same behavior as generating a new segment with a column of all the same values.
IMO second approach will actually cause inconsistent behavior because it will have inverted index with only a single entry, and the segment generated during the reload or from the raw data without a column will be different.
I strongly suggest generating sorted index + dictionary for the default column.

Jackie-Jiang replied:
For the MV default column, since we don't generate sorted index for it, we can consider directly creating inverted index in the default column handler.
Essentially we should try to keep the behavior the same for reloaded segment and segment generated without the data for the default column. That way we don't need any special handling for other index handlers, and the logic will be cleaner.

somandal replied:
@Jackie-Jiang and @siddharthteotia and I discussed these remaining comments and decided to do the following:

- For SV columns the default column should always create a sorted forward index for consistency with the normal segment creation path. We agreed that the query inconsistency that users might see is acceptable since queries may work when they are expected not to work and not the other way around.
- Default multi-value columns should not have a forward index in the long run. For simplifying this PR we decided to ignore the `forwardIndexDisabled` flag for the most part on the reload code path.
- The next change we need to make is address reload code path in general within which we'll add handling for the default column on reload path as well. This will ensure we can use a consistent mechanism to handle the MV default column case above without resorting to special handling in the IndexHandlers.

Jackie-Jiang reviewed (commented):
Mostly good

somandal commented on pinot-tools/src/main/java/org/apache/pinot/tools/segment/converter/DictionaryToRawIndexConverter.java:
```diff
@@ -296,6 +297,10 @@ private void convertOneColumn(IndexSegment segment, String column, File newSegme
     DataSource dataSource = segment.getDataSource(column);
     ForwardIndexReader reader = dataSource.getForwardIndex();
+    if (reader == null) {
+      throw new UnsupportedEncodingException(String.format("Forward index is disabled for column %s, cannot convert!",
```
oops, that was meant to be an `UnsupportedOperationException`. I've modified all of these to use the Preconditions.checkState() instead.

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java:
```diff
@@ -315,6 +341,19 @@ public List<String> getSortedColumns() {
+   * For tests only.
+   */
+  @VisibleForTesting
+  public void setSortedColumns(String sortedColumn) {
```
done

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java:
```diff
@@ -149,7 +154,12 @@ public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, Colum
         }
       }
-      _forwardIndex = indexReaderProvider.newForwardIndexReader(fwdIndexBuffer, metadata);
+      if (forwardIndexDisabled && !metadata.isSorted()) {
+        // Forward index disabled flag is a no-op for sorted columns
+        _forwardIndex = null;
+      } else {
+        _forwardIndex = indexReaderProvider.newForwardIndexReader(fwdIndexBuffer, metadata);
+      }
```
done

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java:
```diff
@@ -133,7 +133,12 @@ public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, Colum
-    PinotDataBuffer fwdIndexBuffer = segmentReader.getIndexFor(columnName, ColumnIndexType.FORWARD_INDEX);
+    // Setting the 'fwdIndexBuffer' to null if forward index is enabled. No-op index readers will be setup for the
+    // forward index disabled columns which doesn't require the 'fwdIndexBuffer'.
+    boolean forwardIndexDisabled = !segmentReader.hasIndexFor(columnName, ColumnIndexType.FORWARD_INDEX);
```
good point, done

somandal commented on pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java:
```diff
@@ -68,9 +68,14 @@ public DataFetcher(Map<String, DataSource> dataSourceMap) {
       DataSource dataSource = entry.getValue();
       DataSourceMetadata dataSourceMetadata = dataSource.getDataSourceMetadata();
+      ForwardIndexReader<?> forwardIndexReader = dataSource.getForwardIndex();
+      if (forwardIndexReader == null) {
+        throw new UnsupportedOperationException(
+            String.format("Forward index disabled for column: %s, cannot create DataFetcher!",
+            dataSourceMetadata.getFieldSpec().getName()));
+      }
```
done - modified this everywhere to use `IllegalStateException`

somandal commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java:
```diff
@@ -58,6 +58,11 @@ public class TableConfig extends BaseJsonConfig {
+  // TODO: Remove this flag once the reload path to create forward index from inverted index and dictionary is
+  //       added. This feature will be disabled until the reload path is updated to handle forward index enable ->
+  //       disable and forward index disable -> enable.
+  public static boolean _disallowForwardIndexDisabled = true;
```
done

somandal commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadataImpl.java:
```diff
@@ -348,6 +348,7 @@ public static class Builder {
     private PartitionFunction _partitionFunction;
     private Set<Integer> _partitions;
     private boolean _autoGenerated;
+    private boolean _forwardIndexDisabled;
```
oops, good point. done

somandal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:
```diff
@@ -897,6 +903,41 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
+      return;
+    }
+
+    boolean forwardIndexDisabled = Boolean.parseBoolean(fieldConfigProperties
+        .getOrDefault(FieldConfig.FORWARD_INDEX_DISABLED, FieldConfig.DEFAULT_FORWARD_INDEX_DISABLED));
```
done

somandal commented:
@siddharthteotia and @Jackie-Jiang can you folks take a look at this PR again? I've addressed the comments we discussed. I've left one test in the `SegmentPreProcesserTest.java` file as a disabled test since we'll be addressing the MV column `forwardIndexDisabled` flag part in Part 2 and I didn't want to rework that test.

amrishlal commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java:
```diff
@@ -402,6 +450,12 @@ private void createDefaultValueColumnV1Indices(String column)
     boolean isSingleValue = fieldSpec.isSingleValueField();
     int maxNumberOfMultiValueElements = isSingleValue ? 0 : 1;
 
+    // TODO: These validations are done as a sanity check though the reload code doesn't actually look at the
```
Does "TODO" refer to items that still need to be resolved in this PR or do they refer to possible future work items that are not part of this PR? Perhaps these comments can be reworded a bit as the "TODO" gives the impression that something has been left in an unresolved state or maybe opening a github issue with a list of future items might be more useful?

somandal replied:
This will be addressed in the next PR to handle reload, I wanted to keep the checks to ensure that someone doesn't enable this feature while using some incompatible configs, but actual handling for default column will be the next related PR. Reworded this comment

amrishlal commented on pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/loader/LoaderTest.java:
```diff
@@ -457,6 +462,147 @@ public void testFSTIndexLoad()
+    indexSegment.destroy();
+
+    // Test scenarios to create a column with no forward index but without enabling inverted index for it
+    assertThrows(IllegalStateException.class, () -> constructSegmentWithForwardIndexDisabled(SegmentVersion.v3, false));
```
We through `IllegalStateException` at lots of places, so it would be good to check the `IllegalStateException` error message as well to make sure that the right `IllegalStateException` was thrown. Same for other places where `assertThrows` is being used.

somandal replied:
done

somandal replied:
I know it looks like a lot of repeats but in a few cases the assertTrue message is a bit different depending on the case we hit. Also wanted to make the Assert.fail() statement a little different depending on the query type to more easily debug. I will need to work on the reload changes for this feature. Perhaps I can revisit improving this in that PR instead if that's okay?

amrishlal commented:
> All code paths that need to read the forward index now have null checks added and appropriate exceptions thrown if the forward index doesn't exist.

> If the forward index is disabled, only a subset of queries can work. Basically queries that don't need to select, transform, group by, order by but do need to use the column in WHERE clauses can be supported.

Can we add compile time checks to catch and fail queries that would not work without forward index being present (or maybe this check was added somewhere and I missed it)? This would avoid wasting resources executing the queries that would fail during runtime due to forward index not being present.

somandal commented:
> > All code paths that need to read the forward index now have null checks added and appropriate exceptions thrown if the forward index doesn't exist.
> 
> > If the forward index is disabled, only a subset of queries can work. Basically queries that don't need to select, transform, group by, order by but do need to use the column in WHERE clauses can be supported.
> 
> Can we add compile time checks to catch and fail queries that would not work without forward index being present (or maybe this check was added somewhere and I missed it)? This would avoid wasting resources executing the queries that would fail during runtime due to forward index not being present.

We considered doing  some kind of static check like this on broker side or planning side but the checks were not straightforward. For e.g. there are many aggregation functions which will work when the forward index is disabled (min, max, some of the distinct count functions) and some which won't (sum, average, etc). We agreed that it's an acceptable tradeoff to fail early during planning for most cases by adding the forward index null checks instead. Few failure scenarios also depend on the existence of other indexes and we'd have to identify such scenarios during execution anyways.

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/operator/filter/ScanBasedFilterOperator.java:
```diff
@@ -43,6 +44,9 @@ public class ScanBasedFilterOperator extends BaseFilterOperator {
     _numDocs = numDocs;
     _nullHandlingEnabled = nullHandlingEnabled;
+    Preconditions.checkState(_dataSource.getForwardIndex() != null,
+        "Forward index disabled for column: %s, creating ScanDocIdSet unsupported!",
```
(nit) a better message may be -- `creating ScanDocIdSet unsupported -> scan based filtering not supported`

somandal replied:
done

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java:
```diff
@@ -133,7 +133,11 @@ public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, Colum
     }
 
-    PinotDataBuffer fwdIndexBuffer = segmentReader.getIndexFor(columnName, ColumnIndexType.FORWARD_INDEX);
+    // Setting the 'fwdIndexBuffer' to null if forward index is enabled. No-op index readers will be setup for the
```
(nit) `typo enabled -> disabled` 

> No-op index readers will be setup for the forward index disabled columns ...

This is no longer the case I guess since we rely on throwing exceptions. So might want to change the comment.

somandal replied:
good catch, done

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:
```diff
@@ -897,6 +903,40 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
+    Preconditions.checkState(indexingConfigs.getInvertedIndexColumns() != null
+            && indexingConfigs.getInvertedIndexColumns().contains(columnName),
+        String.format("Forward index disabled column %s must have inverted index enabled", columnName));
+    Preconditions.checkState(indexingConfigs.getRangeIndexColumns() == null
```
(nit) This check can be cleaned up / broken into 2 checks to make code readable

somandal replied:
done

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:
```diff
@@ -897,6 +903,40 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
+   *     - Validate that either no range index exists for column or the range index version is at least 2 and isn't a
+   *       multi-value column (since mulit-value defaults to index v1).
+   */
+  private static void validateForwardIndexDisabledIndexCompatibility(String columnName, FieldConfig fieldConfig,
```
(nit) we should consider keeping the check either here or during segment generation in `SegmentColumnarIndexCreator`. For now it is fine to safeguard multiple places until feature gets used for a while. 

I am guessing basic validation at the table config level will always be needed since it will come handy even during  enabling / disabling the feature on an existing column. So an unsupported combination will be caught when user attempts to update the table config in controller as opposed to detecting later on the reload path.

siddharthteotia replied:
Ok to keep for now but once we do the reload follow-ups, we should just keep this one because it is just duplicate code / harder to maintain

somandal replied:
yup agree. will definitely remove what we don't need for reload follow ups. Will also try this out in a perf cluster maybe to ensure that just the controller code path is sufficient to catch all cases.

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:
```diff
@@ -897,6 +903,40 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
+    }
+
+    FieldSpec fieldSpec = schema.getFieldSpecFor(columnName);
+    Preconditions.checkState(noDictionaryColumns == null || !noDictionaryColumns.contains(columnName),
```
You should also check that same column has been marked as `RAW` in the `encodingType` in `FieldConfig` since eventually that will be used alone and `noDictionaryColumns` will be deprecated at some point after the migration (hopefully)

siddharthteotia replied:
Looks good.

siddharthteotia commented on pinot-segment-local/src/test/java/org/apache/pinot/segment/local/utils/TableConfigUtilsTest.java:
```diff
@@ -895,6 +895,63 @@ public void testValidateFieldConfig() {
+      tableConfig.setFieldConfigList(Arrays.asList(fieldConfig));
+      TableConfigUtils.validate(tableConfig, schema);
+    } catch (Exception e) {
+      Assert.fail("Should not fail for myCol2 with forward index disabled but is sorted, this is a no-op");
```
Add a test for range index combination scenario as well ?

somandal replied:
done

siddharthteotia commented on pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandlerTest.java:
```diff
@@ -70,7 +70,7 @@ public void setUp()
     _segmentDirectory = new File(_indexDir, driver.getSegmentName());
     _committedSegmentMetadata = new SegmentMetadataImpl(_segmentDirectory);
-    _writer = new SegmentLocalFSDirectory(_indexDir, _committedSegmentMetadata, ReadMode.mmap).createWriter();
+    _writer = new SegmentLocalFSDirectory(_segmentDirectory, _committedSegmentMetadata, ReadMode.mmap).createWriter();
```
not sure why this change is needed

siddharthteotia commented on pinot-segment-local/src/test/resources/data/newColumnsSchemaWithForwardIndexDisabled.json:
```diff
@@ -0,0 +1,77 @@
+      "dataType": "STRING"
+    },
+    {
+      "name": "newForwardIndexDisabledColumnSV",
```
Do we have test for this feature with one of INT / FLOAT / LONG ... other than STRING ?

somandal replied:
Do you want me to add a test in these files for other types? My other query tests use other types. Is that sufficient?

siddharthteotia replied:
As long as we have coverage, I am fine.

somandal replied:
cool then we should be good. as part of reload i can perhaps test out more types on reload path too

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/ForwardIndexDisabledSingleValueQueriesTest.java:
```diff
@@ -0,0 +1,1088 @@
+    }
+    {
+      // Transform function on a selection clause without a forwardIndexDisabled column in transform
+      String query = "SELECT CONCAT(column5, column9, '-') from testTable ORDER BY column1";
```
Can we add test for one nested transform ?

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/ForwardIndexDisabledSingleValueQueriesTest.java:
```diff
@@ -0,0 +1,1088 @@
+    }
+    {
+      // Transform function on a selection clause with a forwardIndexDisabled column in transform
+      String query = "SELECT CONCAT(column6, column9, '-') from testTable";
```
Can we add test for one nested transform ?

siddharthteotia replied:
Also add the transform scenario for WHERE clause, GROUP BY as well

somandal replied:
I have the following transforms covered in the tests:

```
String query = "SELECT column1, column11 from testTable WHERE CONCAT(column6, column9, '-') = '1689277-11270'";

String query = "SELECT column1, column11 from testTable WHERE CONCAT(column5, column9, '-') = 'gFuH-11270' "
          + "ORDER BY column1";

String query = "SELECT CONCAT(column1, column6, '-') FROM testTable GROUP BY CONCAT(column1, column6, '-') "
          + "ORDER BY CONCAT(column1, column6, '-') LIMIT 10";

String query = "SELECT CONCAT(column1, column5, '-') FROM testTable GROUP BY CONCAT(column1, column5, '-') "
          + "ORDER BY CONCAT(column1, column5, '-') LIMIT 10";

String query = "SELECT MAX(ADD(column6, column9)) from testTable LIMIT 10";

String query = "SELECT MAX(ADD(column1, column9)) from testTable LIMIT 10";

String query = "SELECT column1, MAX(ADD(column6, column9)) from testTable GROUP BY column1 LIMIT 10";

String query = "SELECT column1, MAX(ADD(column1, column9)) from testTable GROUP BY column1 LIMIT 10";

String query = "SELECT column1, MAX(ADD(column6, column9)) from testTable GROUP BY column1 ORDER BY column1 "
          + "DESC LIMIT 10";

      String query = "SELECT column1, MAX(ADD(column1, column9)) from testTable GROUP BY column1 ORDER BY column1 "
          + "DESC LIMIT 10";
```

In addition to the above what other transforms do you think I should add? by nested do you mean TRANSFORM inside a TRANSFORM?

somandal replied:
I've added the following, does this cover them all?

```
String query = "SELECT CONCAT(ADD(column1, column6), column5, '-') FROM testTable GROUP BY "
          + "CONCAT(ADD(column1, column6), column5, '-') ORDER BY CONCAT(ADD(column1, column6), column5, '-') LIMIT 10";

      String query = "SELECT CONCAT(ADD(column1, column9), column5, '-') FROM testTable GROUP BY "
          + "CONCAT(ADD(column1, column9), column5, '-') ORDER BY CONCAT(ADD(column1, column9), column5, '-') LIMIT 10";

      String query = "SELECT column1, column11 from testTable WHERE CONCAT(ADD(column6, column1), column9, '-') = "
          + "'1689277-11270'";

      String query = "SELECT column1, column11 from testTable WHERE CONCAT(column5, ADD(column9, column1), '-') = "
          + "'gFuH-2.96708164E8' ORDER BY column1";
```

siddharthteotia replied:
Yep, this is good. Just wanted coverage for transform inside a transform or transform inside an aggregate

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/ForwardIndexDisabledSingleValueQueriesTest.java:
```diff
@@ -0,0 +1,1088 @@
+    }
+    {
+      // Selection query with filters including a column with forwardIndexDisabled enabled on both segments
+      String query = "SELECT column1, column5, column6, column9, column11 FROM testTable WHERE column6 = 2147458029";
```
Can we add a test where `column6` is not in select list but is used in where clause via inverted index `= / IN`

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/ForwardIndexDisabledSingleValueQueriesTest.java:
```diff
@@ -0,0 +1,1088 @@
+    }
+    {
+      // Selection query with filters including a column with forwardIndexDisabled enabled on one segment
+      String query = "SELECT column1, column5, column7, column9, column11 FROM testTable WHERE column7 = 675695";
```
nvm.. looks like this is covered

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/ForwardIndexDisabledSingleValueQueriesTest.java:
```diff
@@ -0,0 +1,1088 @@
+    {
+      // Allowed aggregation functions on forwardIndexDisabled columns
+      String query = "SELECT max(column7), min(column7), count(column7), minmaxrange(column7), "
+          + "distinctcount(column7), distinctcounthll(column6), distinctcountrawhll(column7), "
```
Can you add distinctcountbitmap as well to this list ?

somandal replied:
So `distinctcountbitmap` doesn't work:

```
java.lang.IllegalStateException: Forward index disabled for column: column6, cannot create DataFetcher!

	at com.google.common.base.Preconditions.checkState(Preconditions.java:518)
	at org.apache.pinot.core.common.DataFetcher.<init>(DataFetcher.java:72)
	at org.apache.pinot.core.operator.ProjectionOperator.<init>(ProjectionOperator.java:46)
	at org.apache.pinot.core.plan.ProjectionPlanNode.run(ProjectionPlanNode.java:63)
	at org.apache.pinot.core.plan.TransformPlanNode.run(TransformPlanNode.java:69)
	at org.apache.pinot.core.plan.AggregationPlanNode.buildNonFilteredAggOperator(AggregationPlanNode.java:225)
	at org.apache.pinot.core.plan.AggregationPlanNode.run(AggregationPlanNode.java:80)
	at org.apache.pinot.core.plan.CombinePlanNode.getCombineOperator(CombinePlanNode.java:95)
	at org.apache.pinot.core.plan.CombinePlanNode.run(CombinePlanNode.java:82)
	at org.apache.pinot.core.plan.InstanceResponsePlanNode.run(InstanceResponsePlanNode.java:44)
	at org.apache.pinot.core.plan.GlobalPlanImplV0.execute(GlobalPlanImplV0.java:52)
	at org.apache.pinot.queries.BaseQueriesTest.getBrokerResponse(BaseQueriesTest.java:203)
	at org.apache.pinot.queries.BaseQueriesTest.getBrokerResponse(BaseQueriesTest.java:173)
	at org.apache.pinot.queries.BaseQueriesTest.getBrokerResponse(BaseQueriesTest.java:135)
	at org.apache.pinot.queries.BaseQueriesTest.getBrokerResponse(BaseQueriesTest.java:107)
	at org.apache.pinot.queries.ForwardIndexDisabledSingleValueQueriesTest.testSelectWithAggregationQueries(ForwardIndexDisabledSingleValueQueriesTest.java:807)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:661)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:869)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1193)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:126)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:744)
	at org.testng.TestRunner.run(TestRunner.java:602)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
	at org.testng.TestNG.runSuites(TestNG.java:1144)
	at org.testng.TestNG.run(TestNG.java:1115)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)
```

Do you want me to add a negative test or spend time understanding more about why?

siddharthteotia replied:
Let's make sure to double check this once in follow-up. This PR is now too big to incorporate this here

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/ForwardIndexDisabledSingleValueQueriesTest.java:
```diff
@@ -0,0 +1,1088 @@
+    {
+      // Allowed aggregation functions on non-forwardIndexDisabled columns with group by on non-forwardIndexDisabled
+      // column but order by on allowed aggregation function on forwardIndexDisabled column
+      String query = "SELECT column1, max(column9) from testTable GROUP BY column1 ORDER BY min(column6)";
```
Cover HAVING as well ?

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java:
```diff
@@ -388,8 +416,40 @@ protected boolean createColumnV1Indices(String column)
+   *     - Validate that either no range index exists for column or the range index version is at least 2 and isn't a
+   *       multi-value column (since multi-value defaults to index v1).
+   */
+  protected void validateForwardIndexDisabledConfigsIfPresent(String column, boolean forwardIndexDisabled) {
```
Why is this function needed here ? Table config validator should catch the unsupported combinations / mandatory conditions since anyone who wants to toggle the flag and then leverage reload path has to first update the table config successfully on the controller (which should internally use validator)

somandal replied:
As discussed on the other comment, added this as an additional safeguard. I can remove it though if you think this is unnecessary and the only way is to validate the table configs on the controller. e.g. sometimes we may directly change table config for incompatible schema changes in EI etc. For such cases will the controller catch it for sure?

siddharthteotia replied:
All table config changes must go through controller and it is expected to validate the changed config

siddharthteotia replied:
Ok to do in follow-up PR

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java:
```diff
@@ -363,6 +379,12 @@ protected boolean createColumnV1Indices(String column)
             }
+            // TODO: Support creation of derived columns from forward index disabled columns
+            if (!_segmentWriter.hasIndexFor(argument, ColumnIndexType.FORWARD_INDEX)) {
+              LOGGER.warn("Skip creating derived column: {} because argument: {} does not have a forward index", column,
```
Throwing error is better to force the user to not create a derived column from noForwardIndex column until reload code path supports everything (including derived column handling possibly) ?

somandal replied:
done

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java:
```diff
@@ -224,6 +225,21 @@ Map<String, DefaultColumnAction> computeDefaultColumnActionMap() {
+                "Forward index disabled in segment: " + forwardIndexDisabled + " for column: " + column
+                    + " does not match forward index disabled flag: "
+                    + _indexLoadingConfig.getForwardIndexDisabledColumns().contains(column) + " in the TableConfig, "
+                    + "updating this flag is not supported at the moment.";
```
Not just updating / toggling, even `setting` this flag for the very first time for an existing column is not supported (yet) on the reload path. 

I think we should also log a warning for now and ignore and until the reload support is added, anyone who wants to leverage this feature on an existing column should consider backfilling / refresh

somandal replied:
I've modified the comment but I still think we should fail and throw an exception here. let's discuss more about this if you think a warning log is enough.

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java:
```diff
@@ -402,6 +462,12 @@ private void createDefaultValueColumnV1Indices(String column)
     int maxNumberOfMultiValueElements = isSingleValue ? 0 : 1;
 
+    // Validate that the forwardIndexDisabled flag, if enabled, is compatible with other indexes and configs
+    // For now the forwardIndexDisabled flag is ignored for default columns but will be handled as part of reload
```
(nit) prefix comment at line 466 with TODO

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java:
```diff
@@ -372,6 +394,12 @@ protected boolean createColumnV1Indices(String column)
 
+          // TODO: Support forward index disabled derived column
+          if (_indexLoadingConfig.getForwardIndexDisabledColumns().contains(column)) {
+            LOGGER.warn("Skip creating forward index disabled derived column: {}", column);
```
In this case it is to warn and not throw error since user asked for a feature and we silently ignored (for now). 

But in previous case, error should be thrown imo to tell user to change derived column config's source column or refresh the segments first to generate forward index on source column and then generate target derivedColumn

somandal replied:
sure updated the error message in the previous case to indicate next steps for user

somandal commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java:
```diff
@@ -143,6 +143,16 @@ public TableConfig(@JsonProperty(value = TABLE_NAME_KEY, required = true) String
+      for (FieldConfig fieldConfig : fieldConfigList) {
+        Map<String, String> fieldConfigProperties = fieldConfig.getProperties();
+        if (fieldConfigProperties != null) {
+          boolean forwardIndexDisabled = Boolean.parseBoolean(fieldConfigProperties.getOrDefault(
```
removed this, forgot to clean this up

siddharthteotia commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:
```diff
@@ -897,6 +903,44 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
+    FieldSpec fieldSpec = schema.getFieldSpecFor(columnName);
+    Preconditions.checkState(fieldConfig.getEncodingType() == FieldConfig.EncodingType.DICTIONARY
+            || noDictionaryColumns == null || !noDictionaryColumns.contains(columnName),
+        String.format("Forward index disabled column %s must have dictionary enabled", columnName));
```
@somandal  - this is incorrect. 

Error for this scenario should be thrown if column is marked noDict -- either `encodingType is RAW` or `noDictionaryColumns` contains the column.

This code seems to be throwing error for column marked with DICTIONARY encodingType or if it is not present in `noDictionaryColumns` unless I am misreading the code ?

siddharthteotia replied:
Sorry my bad. I misread the code :) Please ignore

This is preconditions so you are looking for condition to evaluate to TRUE

siddharthteotia commented:
Thanks for the contribution @somandal . Let's please also share the doc on follow-up changes like discussed in last meeting with OSS.
